src/battle_ai_util.c

```diff
@@ -4330,6 +4340,238 @@ bool32 ShouldUseZMove(u32 battlerAtk, u32 battlerDef, u32 chosenMove)
     return FALSE;
 }
 
+void DecideTerastal(u32 battler)
+{
+    if (gBattleStruct->gimmick.usableGimmick[battler] != GIMMICK_TERA) 
+        return;
+    
+
+    if (!(AI_THINKING_STRUCT->aiFlags[battler] & AI_FLAG_SMART_TERA)) 
+        return;
+    
+    // TODO: Currently only single battles are considered.
+    if (IsDoubleBattle())    
+        return; 
+    
+
+    // TODO: A lot of these checks are most effective for an omnicient ai. 
+    // If we don't have enough information about the opponent's moves, consider simpler checks based on type effectivness.
+
+    u32 battlerOpponent = BATTLE_OPPOSITE(battler);
```

Comment:
```suggestion
    u32 opposingBattler = GetOppositeBattler(battler);

```
Renaming to `opposingBattler` isn't strictly necessary it's just the standard we typically use in AI functions so I'd recommend it.

The switch to using `GetOppositeBattler` instead of `BATTLE_OPPOSITE` *is* necessary, as we're actively trying to improve our encapsulation through the battle engine code nowadays.